src/foam/nanos/approval/ApprovableAwareDAO.js

```diff
@@ -0,0 +1,383 @@
+    'foam.core.Detachable',
+    'foam.core.FObject',
+    'foam.core.PropertyInfo',
+    'foam.core.X',
+    'foam.dao.AbstractSink',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.lib.PropertyPredicate',
+    'foam.mlang.MLang',
+    'foam.mlang.MLang.*',
+    'foam.mlang.predicate.Predicate',
+    'foam.nanos.approval.Approvable',
+    'foam.nanos.approval.ApprovableAware',
+    'foam.nanos.approval.ApprovalRequest',
+    'foam.nanos.approval.ApprovalStatus',
+    'foam.nanos.approval.RoleApprovalRequest',
+    'foam.nanos.auth.LifecycleAware',
+    'foam.nanos.auth.LifecycleState',
+    'foam.nanos.auth.User',
+    'foam.nanos.auth.UserQueryService',
+    'foam.nanos.logger.Logger',
+    'foam.nanos.ruler.Operations',
+    'foam.util.SafetyUtil',
+    'java.util.ArrayList',
+    'java.util.Iterator',
+    'java.util.List',
+    'java.util.Map',
+    'java.util.Set',
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'daoKey'
+    },
+    {
+      class: 'Class',
+      name: 'of'
+    },
+    {
+      class: 'Boolean',
+      name: 'isEnabled',
+      value: true
+    }
+  ],
+
+  methods: [
+    {
+      name: 'sendSingleRequest',
+      type: 'void',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'req', type: 'RoleApprovalRequest' },
+        { name: 'userId', type: 'long' }
+      ],
+      javaCode: `
+        RoleApprovalRequest request = (RoleApprovalRequest) req.fclone();
+        request.clearId();
+        request.setApprover(userId);
+        ((DAO) x.get("approvalRequestDAO")).put_(x, request);
+      `
+    },
+    {
+      name: 'fullSend',
+      type: 'void',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'request', type: 'RoleApprovalRequest' },
+        { name: 'obj', type: 'FObject' }
+      ],
+      javaCode:`
+      DAO requestingDAO;
+      Logger logger = (Logger) x.get("logger");
+
+      if ( request.getDaoKey().equals("approvableDAO") ) {
+        DAO approvableDAO = (DAO) x.get("approvableDAO");
+
+        Approvable approvable = (Approvable) approvableDAO.find(request.getObjId());
+
+        requestingDAO = (DAO) x.get(approvable.getDaoKey());
+      } else {
+        requestingDAO = (DAO) x.get(request.getDaoKey());
+      }
+
+      String modelName = requestingDAO.getOf().getObjClass().getSimpleName();
+
+      UserQueryService userQueryService = (UserQueryService) x.get("userQueryService");
+
+      User currentUser = (User) x.get("user");
+
+      List<Long> approverIds = userQueryService.getAllApprovers(getX(), modelName, currentUser);
+
+      if ( approverIds.size() <= 0 ) {
+        logger.error("No Approvers exist for the model: " + modelName);
+        throw new RuntimeException("No Approvers exist for the model: " + modelName);
+      }
+
+      if ( approverIds.size() == 1 && approverIds.get(0) == request.getInitiatingUser() ) {
+        logger.log("The only approver of " + modelName + " is the maker of this request!");
```

Comment:
unique maker/checker should be configuration.   If the approval system is just used for transparency, there is nothing wrong with the maker being the checker.  Was was important is their was a formal recording of some request.

Reply:
Added both canMakerApproveOwnRequest and isTrackingRequestSent